wsmux/stream.go

```diff
@@ -0,0 +1,220 @@
+package wsmux
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+// DefaultCapacity is the maximum length the read buffer will accept
+const DefaultCapacity = 1024
+
+var (
+	// errWriteTimeout is returned if no ack frame arrives before the duration.
+	errWriteTimeout = fmt.Errorf("write operation timed out")
+
+	// errReadTimeout
+	errReadTimeout = fmt.Errorf("read operation timed out")
+
+	// errBufferFull
+	errBufferFull = fmt.Errorf("read buffer is full")
+)
+
+// Stream ...
+type Stream struct {
+	id uint32
+
+	rb       []byte
+	readLock sync.Mutex
+
+	capLock        sync.Mutex
+	remoteCapacity int
+
+	writeChan chan frame
+
+	ackCond  *sync.Cond
+	readCond *sync.Cond
+
+	remoteClosed bool
+	closed       bool
+
+	writeDeadline time.Time
+	readDeadline  time.Time
+
+	session *Session
+}
+
+func newStream(id uint32, writeChan chan frame, session *Session) *Stream {
+	stream := &Stream{
+		id:             id,
+		rb:             make([]byte, 0),
+		writeChan:      writeChan,
+		remoteCapacity: DefaultCapacity,
+		writeDeadline:  time.Now().Add(time.Second * 30),
+		readDeadline:   time.Now().Add(time.Second * 30),
+	}
+	stream.ackCond = sync.NewCond(&stream.capLock)
+	stream.readCond = sync.NewCond(&stream.readLock)
+	stream.session = session
+	return stream
+}
+
+func (s *Stream) ack(read uint32) {
+	defer s.ackCond.Signal()
+	s.capLock.Lock()
+	s.remoteCapacity += int(read)
+	s.capLock.Unlock()
+}
+
+func (s *Stream) write(buf []byte) error {
+	defer s.readCond.Signal()
+	s.readLock.Lock()
+	if len(s.rb)+len(buf) > DefaultCapacity {
+		return errBufferFull
+	}
+	s.rb = append(s.rb, buf...)
+	s.readLock.Unlock()
+	return nil
+}
+
+func (s *Stream) setRemoteClosed() {
+	s.remoteClosed = true
+	if s.closed {
+		s.session.removeStream(s.id)
+	}
+}
+
+func (s *Stream) waitForAck() <-chan struct{} {
+	signal := make(chan struct{}, 1)
```

Comment:
Instead of creating a lot of channels... I'm just thinking could we use a `time.Timer` and then reuse it?
Then we just do `s.writeDeadlineExceeded = true` and `someCond.Broadcast()`...

Just thinking here... maybe it's not prettier... or more efficient...